File: pricing/pricing/report/comprehensive_pricing_analysis/comprehensive_pricing_analysis.js
// Copyright (c) 2024, Frappe Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["Comprehensive Pricing Analysis"] = {
	"filters": [
		{
			"fieldname": "price_lists",
			"label": __("لیست‌های قیمت"),
			"fieldtype": "MultiSelectList",
			"get_data": function(txt) {
				return frappe.db.get_link_options('Price List', txt);
			},
			"reqd": 1
		},
		{
			"fieldname": "price_rounding_amount",
			"label": __("مبلغ رند کردن قیمت"),
			"fieldtype": "Currency",
			"default": 0,
			"description": "قیمت‌ها به این مبلغ به سمت بالا رند می‌شوند"
		},
		{
			"fieldname": "item_group",
			"label": __("گروه کالا"),
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname": "brand",
			"label": __("برند"),
			"fieldtype": "Link",
			"options": "Brand"
		},
		{
			"fieldname": "item_code",
			"label": __("کد کالا"),
			"fieldtype": "Link",
			"options": "Item"
		},
		{
			"fieldname": "enable_installment",
			"label": __("فعال‌سازی قسط"),
			"fieldtype": "Check",
			"default": 0
		},
		{
			"fieldname": "down_payment_percentage",
			"label": __("درصد پیش پرداخت"),
			"fieldtype": "Float",
			"default": 20,
			"depends_on": "eval:doc.enable_installment"
		},
		{
			"fieldname": "number_of_months",
			"label": __("تعداد ماه"),
			"fieldtype": "Int",
			"default": 12,
			"depends_on": "eval:doc.enable_installment"
		},
		{
			"fieldname": "monthly_interest_rate",
			"label": __("نرخ بهره ماهانه (درصد)"),
			"fieldtype": "Float",
			"default": 2,
			"depends_on": "eval:doc.enable_installment"
		},
		{
			"fieldname": "profit_margin",
			"label": __("حاشیه سود (درصد)"),
			"fieldtype": "Float",
			"default": 20,
			"depends_on": "eval:doc.enable_installment"
		},
		{
			"fieldname": "installment_base_price_list",
			"label": __("لیست قیمت مبنای قسط"),
			"fieldtype": "Link",
			"options": "Price List",
			"depends_on": "eval:doc.enable_installment",
			"description": "لیست قیمتی که برای محاسبه قسط استفاده شود"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		if (!value || value === "0" || value === "0.00") {
			return value;
		}
		
		// رنگ‌آمیزی درصد سود
		if (column.fieldname && column.fieldname.includes("profit_pct_")) {
			let numValue = parseFloat(value.toString().replace(/[^\d.-]/g, ''));
			if (!isNaN(numValue)) {
				if (numValue < 0) {
					value = `<span style="color: red; font-weight: bold;">${value}</span>`;
				} else if (numValue > 30) {
					value = `<span style="color: green; font-weight: bold;">${value}</span>`;
				} else if (numValue > 15) {
					value = `<span style="color: blue;">${value}</span>`;
				}
			}
		}
		
		// رنگ‌آمیزی سود
		if (column.fieldname && column.fieldname.includes("profit_") && !column.fieldname.includes("profit_pct_")) {
			let numValue = parseFloat(value.toString().replace(/[^\d.-]/g, ''));
			if (!isNaN(numValue)) {
				if (numValue < 0) {
					value = `<span style="color: red;">${value}</span>`;
				} else if (numValue > 0) {
					value = `<span style="color: green;">${value}</span>`;
				}
			}
		}
		
		// رنگ‌آمیزی قیمت قسطی
		if (column.fieldname == "installment_price") {
			value = `<span style="color: purple; font-weight: bold;">${value}</span>`;
		}
		
		// رنگ‌آمیزی بهره
		if (column.fieldname == "total_interest") {
			let numValue = parseFloat(value.toString().replace(/[^\d.-]/g, ''));
			if (!isNaN(numValue) && numValue > 0) {
				value = `<span style="color: orange;">${value}</span>`;
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// اضافه کردن دکمه‌های اضافی
		report.page.add_inner_button(__("صادرات به Excel"), function() {
			let data = frappe.query_report.data || [];
			let columns = frappe.query_report.columns || [];
			
			if (data.length === 0) {
				frappe.msgprint(__("هیچ داده‌ای برای صادرات وجود ندارد"));
				return;
			}
			
			// تبدیل داده‌ها به فرمت CSV
			let csv_data = [];
			let headers = columns.map(col => col.label || col.fieldname);
			csv_data.push(headers);
			
			data.forEach(row => {
				let row_data = columns.map(col => row[col.fieldname] || '');
				csv_data.push(row_data);
			});
			
			frappe.tools.downloadify(csv_data, null, "تحلیل_جامع_قیمت‌گذاری");
		});
		
		report.page.add_inner_button(__("چاپ گزارش"), function() {
			if (!frappe.query_report.data || frappe.query_report.data.length === 0) {
				frappe.msgprint(__("هیچ داده‌ای برای چاپ وجود ندارد"));
				return;
			}
			window.print();
		});
		
		// اضافه کردن دکمه تنظیمات سریع
		report.page.add_inner_button(__("تنظیمات سریع"), function() {
			let d = new frappe.ui.Dialog({
				title: 'تنظیمات سریع گزارش',
				fields: [
					{
						label: 'نوع تحلیل',
						fieldname: 'analysis_type',
						fieldtype: 'Select',
						options: [
							'',
							'تحلیل سودآوری',
							'مقایسه قیمت‌ها',
							'تحلیل قسطی'
						]
					},
					{
						label: 'رند کردن قیمت',
						fieldname: 'quick_rounding',
						fieldtype: 'Currency',
						description: 'مبلغ رند کردن قیمت (مثال: 1000)'
					}
				],
				primary_action_label: 'اعمال',
				primary_action(values) {
					if (values.analysis_type === 'تحلیل قسطی') {
						frappe.query_report.set_filter_value('enable_installment', 1);
					}
					if (values.quick_rounding) {
						frappe.query_report.set_filter_value('price_rounding_amount', values.quick_rounding);
					}
					frappe.query_report.refresh();
					d.hide();
				}
			});
			d.show();
		});
		
		// راهنمای استفاده
		report.page.add_inner_button(__("راهنما"), function() {
			frappe.msgprint({
				title: 'راهنمای استفاده از گزارش',
				message: `
					<div style="text-align: right; direction: rtl;">
						<h4>نحوه استفاده:</h4>
						<ul>
							<li><strong>لیست‌های قیمت:</strong> نام لیست‌های قیمت را با کاما جدا کنید</li>
							<li><strong>تحلیل قسطی:</strong> برای فعال‌سازی محاسبات قسط، گزینه "فعال‌سازی قسط" را انتخاب کنید</li>
							<li><strong>فیلترها:</strong> از فیلترهای گروه کالا و برند برای محدود کردن نتایج استفاده کنید</li>
						</ul>
						<h4>رنگ‌بندی:</h4>
						<ul>
							<li><span style="color: green;">سبز:</span> سود بالا (بیش از 30%)</li>
							<li><span style="color: blue;">آبی:</span> سود متوسط (15-30%)</li>
							<li><span style="color: red;">قرمز:</span> ضرر یا سود کم</li>
							<li><span style="color: purple;">بنفش:</span> قیمت قسطی</li>
							<li><span style="color: orange;">نارنجی:</span> بهره</li>
						</ul>
					</div>
				`,
				indicator: 'blue'
			});
		});
	}
};

File: pricing/pricing/report/comprehensive_pricing_analysis/comprehensive_pricing_analysis.py
# Copyright (c) 2024, Frappe Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt, cint
import math

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data

def get_columns(filters):
    columns = [
        {
            "label": "کد کالا",
            "fieldname": "item_code",
            "fieldtype": "Link",
            "options": "Item",
            "width": 120
        },
        {
            "label": "نام کالا",
            "fieldname": "item_name",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": "گروه کالا",
            "fieldname": "item_group",
            "fieldtype": "Link",
            "options": "Item Group",
            "width": 120
        }
    ]
    
    # اضافه کردن ستون‌های لیست قیمت‌های انتخاب شده
    if filters.get("price_lists"):
        if isinstance(filters.get("price_lists"), list):
            price_lists = filters.get("price_lists")
        else:
            price_lists = filters.get("price_lists").split(",")
        for price_list in price_lists:
            price_list = price_list.strip()
            columns.extend([
                {
                    "label": f"قیمت - {price_list}",
                    "fieldname": f"price_{price_list.replace(' ', '_')}",
                    "fieldtype": "Currency",
                    "width": 120
                },
                {
                    "label": f"سود - {price_list}",
                    "fieldname": f"profit_{price_list.replace(' ', '_')}",
                    "fieldtype": "Currency",
                    "width": 100
                },
                {
                    "label": f"درصد سود - {price_list}",
                    "fieldname": f"profit_pct_{price_list.replace(' ', '_')}",
                    "fieldtype": "Percent",
                    "width": 100
                }
            ])
    
    # اضافه کردن ستون‌های هزینه
    columns.extend([
        {
            "label": "هزینه مواد اولیه",
            "fieldname": "raw_material_cost",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": "هزینه برق",
            "fieldname": "electricity_cost",
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "label": "هزینه مصرفی",
            "fieldname": "consumable_cost",
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "label": "هزینه اجاره",
            "fieldname": "rent_cost",
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "label": "هزینه نیروی کار",
            "fieldname": "labor_cost",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": "هزینه عملیات",
            "fieldname": "operation_cost",
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "label": "هزینه سربار",
            "fieldname": "overhead_cost",
            "fieldtype": "Currency",
            "width": 100
        },
        {
            "label": "مجموع هزینه",
            "fieldname": "total_cost",
            "fieldtype": "Currency",
            "width": 120
        }
    ])
    
    # اضافه کردن ستون قیمت قسطی اگر فعال باشد
    if filters.get("enable_installment"):
        columns.extend([
            {
                "label": "قیمت با بهره",
                "fieldname": "installment_price",
                "fieldtype": "Currency",
                "width": 120
            },
            {
                "label": "پیش پرداخت",
                "fieldname": "down_payment",
                "fieldtype": "Currency",
                "width": 100
            },
            {
                "label": "قسط ماهانه",
                "fieldname": "monthly_payment",
                "fieldtype": "Currency",
                "width": 100
            },
            {
                "label": "مجموع بهره",
                "fieldname": "total_interest",
                "fieldtype": "Currency",
                "width": 100
            }
        ])
    
    return columns

def get_data(filters):
    conditions = get_conditions(filters)
    
    # دریافت کالاها
    items_query = f"""
        SELECT 
            i.item_code,
            i.item_name,
            i.item_group,
            i.standard_rate
        FROM `tabItem` i
        WHERE i.disabled = 0
        {conditions}
        ORDER BY i.item_code
    """
    
    items = frappe.db.sql(items_query, filters, as_dict=1)
    
    data = []
    for item in items:
        row = {
            "item_code": item.item_code,
            "item_name": item.item_name,
            "item_group": item.item_group
        }
        
        # محاسبه هزینه‌ها از BOM
        cost_data = get_item_costs(item.item_code)
        row.update(cost_data)
        
        # دریافت قیمت‌ها از لیست‌های قیمت انتخاب شده
        if filters.get("price_lists"):
            if isinstance(filters.get("price_lists"), list):
                price_lists = filters.get("price_lists")
            else:
                price_lists = filters.get("price_lists").split(",")
            for price_list in price_lists:
                price_list = price_list.strip()
                price = get_item_price(item.item_code, price_list)
                
                field_suffix = price_list.replace(' ', '_')
                row[f"price_{field_suffix}"] = price
                
                # اعمال رند کردن قیمت
                if filters.get("price_rounding_amount") and price > 0:
                    rounding_amount = flt(filters.get("price_rounding_amount"))
                    if rounding_amount > 0:
                        price = math.ceil(price / rounding_amount) * rounding_amount
                        row[f"price_{field_suffix}"] = price
                
                # محاسبه سود
                if price and row.get("total_cost"):
                    profit = price - row["total_cost"]
                    profit_pct = (profit / row["total_cost"]) * 100 if row["total_cost"] > 0 else 0
                    row[f"profit_{field_suffix}"] = profit
                    row[f"profit_pct_{field_suffix}"] = profit_pct
        
        # محاسبه قیمت قسطی
        if filters.get("enable_installment"):
            installment_data = calculate_installment_price(
                row.get("total_cost", 0),
                filters
            )
            row.update(installment_data)
        
        data.append(row)
    
    return data

def get_conditions(filters):
    conditions = ""
    
    if filters.get("item_group"):
        conditions += " AND i.item_group = %(item_group)s"
    
    if filters.get("brand"):
        conditions += " AND i.brand = %(brand)s"
    
    if filters.get("item_code"):
        conditions += " AND i.item_code = %(item_code)s"
    
    return conditions

def get_item_costs(item_code):
    """محاسبه هزینه‌های کالا از BOM"""
    
    # دریافت BOM فعال
    bom = frappe.db.get_value("BOM", {
        "item": item_code,
        "is_active": 1,
        "is_default": 1
    }, ["name", "raw_material_cost", "operating_cost"], as_dict=1)
    
    if not bom:
        return {
            "raw_material_cost": 0,
            "electricity_cost": 0,
            "consumable_cost": 0,
            "rent_cost": 0,
            "labor_cost": 0,
            "operation_cost": 0,
            "overhead_cost": 0,
            "total_cost": 0
        }
    
    # محاسبه هزینه مواد اولیه
    raw_material_cost = flt(bom.raw_material_cost)
    
    # محاسبه هزینه عملیات
    operation_cost = flt(bom.operating_cost)
    
    # محاسبه سایر هزینه‌ها (می‌توان از تنظیمات یا جداول مخصوص دریافت کرد)
    electricity_cost = raw_material_cost * 0.05  # 5% از مواد اولیه
    consumable_cost = raw_material_cost * 0.03   # 3% از مواد اولیه
    rent_cost = raw_material_cost * 0.02         # 2% از مواد اولیه
    labor_cost = operation_cost * 0.6            # 60% از هزینه عملیات
    overhead_cost = (raw_material_cost + operation_cost) * 0.1  # 10% سربار
    
    total_cost = (raw_material_cost + electricity_cost + consumable_cost + 
                 rent_cost + labor_cost + operation_cost + overhead_cost)
    
    return {
        "raw_material_cost": raw_material_cost,
        "electricity_cost": electricity_cost,
        "consumable_cost": consumable_cost,
        "rent_cost": rent_cost,
        "labor_cost": labor_cost,
        "operation_cost": operation_cost,
        "overhead_cost": overhead_cost,
        "total_cost": total_cost
    }

def get_item_price(item_code, price_list):
    """دریافت قیمت کالا از لیست قیمت"""
    price = frappe.db.get_value("Item Price", {
        "item_code": item_code,
        "price_list": price_list
    }, "price_list_rate")
    
    return flt(price) if price else 0

def calculate_installment_price(base_cost, filters):
    """محاسبه قیمت قسطی با بهره"""
    
    if not filters.get("enable_installment"):
        return {}
    
    # محاسبه قیمت فروش با حاشیه سود
    profit_margin = flt(filters.get("profit_margin", 20))  # 20% پیش‌فرض
    selling_price = base_cost * (1 + profit_margin / 100)
    
    # پارامترهای قسط
    down_payment_pct = flt(filters.get("down_payment_percentage", 20))
    months = cint(filters.get("number_of_months", 12))
    monthly_rate = flt(filters.get("monthly_interest_rate", 2)) / 100
    
    # محاسبه پیش پرداخت
    down_payment = selling_price * (down_payment_pct / 100)
    financed_amount = selling_price - down_payment
    
    if months <= 0 or monthly_rate <= 0:
        return {
            "installment_price": selling_price,
            "down_payment": down_payment,
            "monthly_payment": financed_amount / months if months > 0 else 0,
            "total_interest": 0
        }
    
    # محاسبه قسط ماهانه با بهره مرکب
    monthly_payment = financed_amount * (monthly_rate * (1 + monthly_rate)**months) / ((1 + monthly_rate)**months - 1)
    total_payments = monthly_payment * months
    total_amount = down_payment + total_payments
    total_interest = total_amount - selling_price
    
    return {
        "installment_price": total_amount,
        "down_payment": down_payment,
        "monthly_payment": monthly_payment,
        "total_interest": total_interest
    }
